ComponentKit/Utilities/CKComponentAction.h

```diff
@@ -11,44 +11,102 @@
   /** Starts searching at the sender's next responder. Usually this is what you want to prevent infinite loops. */
   CKComponentActionSendBehaviorStartAtSenderNextResponder,
   /** If the sender itself responds to the action, invoke the action on the sender. */
   CKComponentActionSendBehaviorStartAtSender,
 };
 
+template<typename... T>
+struct CKTypedComponentAction {
+  CKTypedComponentAction() : _variant(CKTypedComponentActionVariantRawSelector), _target(nil), _selector(NULL) {};
+  CKTypedComponentAction(id target, SEL selector) : _variant(CKTypedComponentActionVariantTargetSelector), _target(target), _selector(selector)
+  {
+#if DEBUG
+    std::vector<const char *> typeEncodings;
+    _CKTypedComponentActionTypeVectorBuild(typeEncodings, _CKTypedComponentActionTypelist<T...>{});
+    _CKTypedComponentDebugCheckTargetSelector(target, selector, typeEncodings);
+#endif
+  }
+  
+  /** Legacy constructor for raw selector actions. Traverse up the mount responder chain. */
+  CKTypedComponentAction(SEL selector) : _variant(CKTypedComponentActionVariantRawSelector), _target(nil), _selector(selector) { };
+  
+  explicit operator bool() const { return _selector != NULL; };
+  bool operator==(const CKTypedComponentAction& rhs) const
+  {
+    return _selector == rhs._selector && CKObjectIsEqual(_target, rhs._target) && _variant == rhs._variant;
+  }
+  
+  SEL selector() const { return _selector; };
+  
+  void send(CKComponent *sender, T... args) const
+  { this->send(sender, CKComponentActionSendBehaviorStartAtSenderNextResponder, args...); }
+  void send(CKComponent *sender, CKComponentActionSendBehavior behavior, T... args) const
+  {
+    if (_selector == NULL) {
+      return;
+    }
+    const id target = _variant == CKTypedComponentActionVariantRawSelector ? sender : _target;
+    const id initialTarget = behavior == CKComponentActionSendBehaviorStartAtSender ? target : [target nextResponder];
+    _CKComponentActionSendResponderChain(_selector, initialTarget, sender, args...);
+  }
+  
+  /** Allows you to get a block that sends the action when executed. */
+  typedef void (^CKTypedComponentActionExecutionBlock)(id sender, T... args);
+  CKTypedComponentActionExecutionBlock block() const;
+  
+  /** We support promotion from actions that take no arguments, but we do not allow demotion. */
+  CKTypedComponentAction(const CKTypedComponentAction<> &action)
+  { _variant = action._variant; _target = action._target; _selector = action._selector; }
+
+  /** 
+   We allow demotion from actions with types to untyped actions, but only when explicit. This means arguments to the
+   method specified here will have nil values at runtime. Used for interoperation with older API's.
+   */
+  template<typename... Ts>
+  explicit CKTypedComponentAction<>(const CKTypedComponentAction<Ts...> &action)
+  { _variant = action._variant; _target = action._target; _selector = action._selector; }
+  
+  /** Allows conversion from NULL actions. */
```

Comment:
This is just for conversion. Ideally I'd be able to strip these out and you can use {} instead. Only problem with that is that {} can't be used on the right side of a ternary... so it'll inflate the size of the call a bit if we did this.